### PR TITLE
[Merged by Bors] - feat(Data/Real/EReal): add inverse and division of `ENNReal`s

### PR DESCRIPTION
**Pull request by pitmonticone**
This PR adds API for inverse and division  of extended reals. 

Co-authored-by: @D-Thomine 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by sgouezel — COMMENTED** [review]
It's a bad idea to define a non-homogeneous division on EReal → ENNReal → EReal. Instead, as I discussed with Damien, one should define an homogeneous division on EReal, first be defining an inverse function and then using (a, b) -> a * b⁻¹.

**Comment by pitmonticone** [comment]
Thank you @sgouezel. I didn't know that. 

Is it ok now?

**Review by sgouezel — COMMENTED** [review]
Did you check with Damien if that corresponds to what he wants to do? In particular, did you check with him that the API still corresponds to what he needs for the topological entropy? Again, there's no rush, but you should make sure things are done right (for instance, there a lot of `induction'` in the PR, while I told you in a previous one that you should favor `induction` :-)

And awesome work, by the way!

**Comment by pitmonticone** [comment]
Yes, I believe that is the new API in BET/TopologicalEntropy. 

@D-Thomine can you confirm?

File: Mathlib/Data/Real/EReal.lean [hunk 1]
```diff
@@ -1471,6 +1471,257 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+  | top_pos x x_pos => rw [top_mul_of_pos (EReal.coe_pos.2 x_pos), inv_top, zero_mul]
+  | top_neg x x_neg => rw [top_mul_of_neg (EReal.coe_neg'.2 x_neg), inv_bot, inv_top, zero_mul]
+  | pos_bot x x_pos => rw [mul_bot_of_pos (EReal.coe_pos.2 x_pos), inv_bot, mul_zero]
+  | coe_coe x y => rw [← coe_mul, ← coe_inv, _root_.mul_inv, coe_mul, coe_inv, coe_inv]
+  | neg_bot x x_neg => rw [mul_bot_of_neg (EReal.coe_neg'.2 x_neg), inv_top, inv_bot, mul_zero]
+
+/-! #### Inversion and Absolute Value -/
+
+theorem sign_mul_inv_abs (a : EReal) : (sign a) * (a.abs : EReal)⁻¹ = a⁻¹ := by
+  induction' a using EReal.rec with a
```
[the remark this answers is not in the document]
Comment (pitmonticone) [reply Mathlib/Data/Real/EReal.lean]: Sure. Done. Thanks.

File: Mathlib/Data/Real/EReal.lean [hunk 2]
```diff
@@ -1471,6 +1471,257 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
 #align ereal.coe_ennreal_pow EReal.coe_ennreal_pow
 
+/-! ### Inverse -/
+
+noncomputable instance : DivInvMonoid EReal where
+  inv := fun (a : EReal) ↦ match a with
+  | none => 0
+  | some b => match b with
+    | none => 0
+    | some c => (c⁻¹ : ℝ)
```
[the remark this answers is not in the document]
Comment (pitmonticone) [reply Mathlib/Data/Real/EReal.lean]: Done, thanks.

File: Mathlib/Data/Real/EReal.lean [hunk 3]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
   map_pow (⟨⟨(↑), coe_ennreal_one⟩, coe_ennreal_mul⟩ : ℝ≥0∞ →* EReal) _ _
 #align ereal.coe_ennreal_pow EReal.coe_ennreal_pow
 
+/-! ### Inverse -/
+
+/-- Mmultiplicative inverse of an `EReal`. -/
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1476]:
```suggestion
/-- Multiplicative inverse of an `EReal`. We choose `0⁻¹ = 0` to guarantee several good properties, for
instance `(a * b)⁻¹ = a⁻¹ * b⁻¹`. -/
```

File: Mathlib/Data/Real/EReal.lean [hunk 4]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+  induction a using EReal.rec with
+  | h_bot | h_top => simp
+  | h_real a => rw [← coe_inv a, EReal.coe_nonneg, inv_nonneg]; exact EReal.coe_nonneg.1 h
+
+theorem inv_nonpos_of_nonpos {a : EReal} (h : a ≤ 0) : a⁻¹ ≤ 0 := by
+  induction a using EReal.rec with
+  | h_bot | h_top => simp
+  | h_real a => rw [← coe_inv a, EReal.coe_nonpos, inv_nonpos]; exact EReal.coe_nonpos.1 h
+
+theorem inv_pos_of_ntop_pos {a : EReal} (h : 0 < a) (h' : a ≠ ⊤) : 0 < a⁻¹ := by
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1565]:
```suggestion
theorem inv_pos_of_pos_ne_top {a : EReal} (h : 0 < a) (h' : a ≠ ⊤) : 0 < a⁻¹ := by
```
To follow the order of symbols in the statement, and also `ntop` is not a standard naming convention.

File: Mathlib/Data/Real/EReal.lean [hunk 5]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+  induction a using EReal.rec with
+  | h_bot => exact (not_lt_bot h).rec
+  | h_real a =>  rw [← coe_inv a]; norm_cast at *; exact inv_pos_of_pos h
+  | h_top => exact (h' (Eq.refl ⊤)).rec
+
+theorem inv_neg_of_nbot_neg {a : EReal} (h : a < 0) (h' : a ≠ ⊥) : a⁻¹ < 0 := by
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1571]:
```suggestion
theorem inv_neg_of_neg_ne_bot {a : EReal} (h : a < 0) (h' : a ≠ ⊥) : a⁻¹ < 0 := by
```

File: Mathlib/Data/Real/EReal.lean [hunk 6]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+
+@[simp]
+theorem div_zero {a : EReal} : a / 0 = 0 := by
+  change a * 0⁻¹ = 0
+  rw [inv_zero, mul_zero a]
+
+@[simp]
+theorem zero_div {a : EReal} : 0 / a = 0 := zero_mul a⁻¹
+
+theorem top_div_nontop_pos {a : EReal} (h : 0 < a) (h' : a ≠ ⊤) : ⊤ / a = ⊤ :=
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1597]:
```suggestion
theorem top_div_of_pos_ne_top {a : EReal} (h : 0 < a) (h' : a ≠ ⊤) : ⊤ / a = ⊤ :=
```

File: Mathlib/Data/Real/EReal.lean [hunk 7]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+  top_mul_of_pos (inv_pos_of_ntop_pos h h')
+
+theorem top_div_nonbot_neg {a : EReal} (h : a < 0) (h' : a ≠ ⊥) : ⊤ / a = ⊥ :=
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1600]:
```suggestion
theorem top_div_of_neg_ne_bot {a : EReal} (h : a < 0) (h' : a ≠ ⊥) : ⊤ / a = ⊥ :=
```

File: Mathlib/Data/Real/EReal.lean [hunk 8]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+  top_mul_of_neg (inv_neg_of_nbot_neg h h')
+
+theorem bot_div_nontop_pos {a : EReal} (h : 0 < a) (h' : a ≠ ⊤) : ⊥ / a = ⊥ :=
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1603]:
```suggestion
theorem bot_div_of_pos_ne_top {a : EReal} (h : 0 < a) (h' : a ≠ ⊤) : ⊥ / a = ⊥ :=
```

File: Mathlib/Data/Real/EReal.lean [hunk 9]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+  bot_mul_of_pos (inv_pos_of_ntop_pos h h')
+
+theorem bot_div_nonbot_neg {a : EReal} (h : a < 0) (h' : a ≠ ⊥) : ⊥ / a = ⊤ :=
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1606]:
```suggestion
theorem bot_div_of_neg_ne_bot {a : EReal} (h : a < 0) (h' : a ≠ ⊥) : ⊥ / a = ⊤ :=
```

File: Mathlib/Data/Real/EReal.lean [hunk 10]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+theorem mul_div_mul_cancel {a b c : EReal} (h₁ : c ≠ ⊥) (h₂ : c ≠ ⊤) (h₃ : c ≠ 0) :
+    (a * c) / (b * c) = a / b := by
+  change (a * c) * (b * c)⁻¹ = a * b⁻¹
+  rw [mul_assoc, mul_inv b c]
+  congr
+  exact mul_div_cancel h₁ h₂ h₃
+
+/-! #### Division Distributivity -/
+
+theorem div_right_distrib_of_nneg {a b c : EReal} (h : 0 ≤ a) (h' : 0 ≤ b) :
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1644]:
```suggestion
theorem div_right_distrib_of_nonneg {a b c : EReal} (h : 0 ≤ a) (h' : 0 ≤ b) :
```

File: Mathlib/Data/Real/EReal.lean [hunk 11]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+    (a + b) / c = (a / c) + (b / c) :=
+  EReal.right_distrib_of_nonneg h h'
+
+/-! #### Division and Order s-/
+
+theorem monotone_div_right_of_nonneg {b : EReal} (h : 0 ≤ b) : Monotone fun a ↦ a / b :=
+  fun _ _ h' ↦ mul_le_mul_of_nonneg_right h' (inv_nonneg_of_nonneg h)
+
+theorem monotone_div_right_of_nonneg_apply {a a' b : EReal} (h : 0 ≤ b) (h' : a ≤ a') :
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1653]:
```suggestion
theorem div_le_div_right_of_nonneg {a a' b : EReal} (h : 0 ≤ b) (h' : a ≤ a') :
```

File: Mathlib/Data/Real/EReal.lean [hunk 12]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+theorem strictMono_div_right_of_pos {b : EReal} (h : 0 < b) (h' : b ≠ ⊤) :
+    StrictMono fun a ↦ a / b := by
+  intro a a' a_lt_a'
+  apply lt_of_le_of_ne <| monotone_div_right_of_nonneg_apply (le_of_lt h) (le_of_lt a_lt_a')
+  intro hyp
+  apply ne_of_lt a_lt_a'
+  rw [← @EReal.mul_div_cancel a b (ne_bot_of_gt h) h' (ne_of_gt h), hyp,
+    @EReal.mul_div_cancel a' b (ne_bot_of_gt h) h' (ne_of_gt h)]
+
+theorem strictMono_div_right_of_pos_apply {a a' b : EReal} (h₁ : 0 < b) (h₂ : b ≠ ⊤)
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1666]:
```suggestion
theorem div_lt_div_right_of_pos {a a' b : EReal} (h₁ : 0 < b) (h₂ : b ≠ ⊤)
```

File: Mathlib/Data/Real/EReal.lean [hunk 13]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+
+theorem antitone_div_right_of_nonpos {b : EReal} (h : b ≤ 0) : Antitone fun a ↦ a / b := by
+  intro a a' h'
+  change a' * b⁻¹ ≤ a * b⁻¹
+  rw [← neg_neg (a * b⁻¹), ← neg_neg (a' * b⁻¹), neg_le_neg_iff, mul_comm a b⁻¹, mul_comm a' b⁻¹,
+    ← neg_mul b⁻¹ a, ← neg_mul b⁻¹ a', mul_comm (-b⁻¹) a, mul_comm (-b⁻¹) a', ← inv_neg b]
+  have : 0 ≤ -b := by apply le_neg_of_le_neg; simp [h]
+  exact monotone_div_right_of_nonneg_apply this h'
+
+theorem antitone_div_right_of_nonpos_apply {a a' b : EReal} (h : b ≤ 0) (h' : a ≤ a') :
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1678]:
```suggestion
theorem div_le_div_right_of_nonpos {a a' b : EReal} (h : b ≤ 0) (h' : a ≤ a') :
```

File: Mathlib/Data/Real/EReal.lean [hunk 14]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+    StrictAnti fun a ↦ a / b := by
+  intro a a' a_lt_a'
+  simp only
+  apply lt_of_le_of_ne <| antitone_div_right_of_nonpos_apply (le_of_lt h) (le_of_lt a_lt_a')
+  intro hyp
+  apply ne_of_lt a_lt_a'
+  rw [← @EReal.mul_div_cancel a b h' (ne_top_of_lt h) (ne_of_lt h), ← hyp,
+    @EReal.mul_div_cancel a' b h' (ne_top_of_lt h) (ne_of_lt h)]
+
+theorem strictAnti_div_right_of_neg_apply {a a' b : EReal} (h₁ : b < 0) (h₂ : b ≠ ⊥)
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1692]:
```suggestion
theorem div_lt_div_right_of_neg {a a' b : EReal} (h₁ : b < 0) (h₂ : b ≠ ⊥)
```

File: Mathlib/Data/Real/EReal.lean [hunk 15]
```diff
@@ -1471,6 +1471,252 @@ theorem coe_ennreal_pow (x : ℝ≥0∞) (n : ℕ) : (↑(x ^ n) : EReal) = (x :
...
+    (h₃ : a < a') : a' / b < a / b :=
+  strictAnti_div_right_of_neg h₁ h₂ h₃
+
+theorem le_div_iff_mul_le {a b c : EReal} (h : b > 0) (h' : b ≠ ⊤) :
+    a ≤ c / b ↔ a * b ≤ c := by
+  nth_rw 1 [← @mul_div_cancel a b (ne_bot_of_gt h) h' (ne_of_gt h)]
+  rw [mul_div b a b, mul_comm a b]
+  exact StrictMono.le_iff_le (strictMono_div_right_of_pos h h')
+
+theorem div_le_iff_le_mul {a b c : EReal} (h : b > 0) (h' : b ≠ ⊤) :
```
Comment (sgouezel) [inline Mathlib/Data/Real/EReal.lean:1702]:
```suggestion
theorem div_le_iff_le_mul {a b c : EReal} (h : 0 < b) (h' : b ≠ ⊤) :
```

**Review by sgouezel — COMMENTED** [review]
This looks very good. Minor tweaks are required about the naming of lemmas, though.

**Comment by pitmonticone** [comment]
I should have addressed all review comments. 

Thank you very much @sgouezel for your review!

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**